tests/integration/aliases/step2/index.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright 2016-2018, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+
+const stackName = pulumi.getStack();
+const projectName = pulumi.getProject();
+
+class Resource extends pulumi.ComponentResource {
+    constructor(name: string, opts?: pulumi.ComponentResourceOptions) {
+        super("my:module:Resource", name, {}, opts);
+    }
+}
+
+// Scenario #1 - rename a resource
+// This resource was previously named `one`, we'll alias to the old name.
+const res1 = new Resource("newres1", {
+    aliases: [`urn:pulumi:${stackName}::${projectName}::my:module:Resource::res1`],
+});
+
+// Scenario #2 - adopt a resource into a component The component author is the same as the component user, and changes
+// the component to be able to adopt the resource that was previously defined separately...
+class Component extends pulumi.ComponentResource {
+    resource: Resource;
+    constructor(name: string, opts?: pulumi.ComponentResourceOptions) {
+        super("my:module:Component", name, {}, opts);
+        // The resource creation was moved from top level to inside the component.
+        this.resource = new Resource(`${name}-child`, {
+            // With a new parent
+            parent: this,
+            // But with an alias provided based on knowing where the resource existing before - in this case at top
+            // level.
+            aliases: [`urn:pulumi:${stackName}::${projectName}::my:module:Resource::res2`],
+        });
+    }
+}
+// The creation of the component is unchanged.
+const comp2 = new Component("comp2");
+
+// Scenario #3 - rename a component (and all it's children)
+// No change to the component...
+class ComponentThree extends pulumi.ComponentResource {
+    resource1: Resource;
+    resource2: Resource;
+    constructor(name: string, opts?: pulumi.ComponentResourceOptions) {
+        super("my:module:ComponentThree", name, {}, opts);
+        // Note that both un-prefixed and parent-name-prefixed child names are supported. For the later, the implicit
+        // alias inherited from the parent alias will include replacing the name prefix to match the parent alias name.
+        this.resource1 = new Resource(`${name}-child`, {parent: this});
+        this.resource2 = new Resource("otherchild", {parent: this});
+    }
+}
+// ...but applying an alias to the instance succesfully renames both the component and the children.
+const comp3 = new ComponentThree("newcomp3", {
+    aliases: [`urn:pulumi:${stackName}::${projectName}::my:module:ComponentThree::comp3`],
+});
+
+// Scenario #4 - change the type of a component
+class ComponentFour extends pulumi.ComponentResource {
+    resource: Resource;
+    constructor(name: string, opts?: pulumi.ComponentResourceOptions) {
+        const aliases = (opts && opts.aliases) || [];
+        aliases.push(`urn:pulumi:${stackName}::${projectName}::my:module:ComponentFour::${name}`);
```

Comment:
Do you think we should just expose `createTestUrn` (renaming it to `createUrn` or something) so folks don't have to do this ``urn:pulumi:${stackName}::${projectName}::` dance each time?

I see this is point three above, I think we should just do it.

Reply:
Yes - I strongly expect we will want to publicly expose the `childURN` function at https://github.com/pulumi/pulumi/pull/2404/files#diff-60a0d91ede7b9f2ccd5436be151b9057R29.  (We can then use that in place of `createTestUrn` as well).  

I *think* that that one function `createUrn(parent: URN | Resource | undefined, type: string, name: string) ` is sufficient for all the scenarios here.